core-primitives/extrinsics-factory/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(all(feature = "std", feature = "sgx"))]
+compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+extern crate sgx_tstd as std;
+
+// re-export module to properly feature gate sgx and regular std environment
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+pub mod sgx_reexport_prelude {
+	pub use thiserror_sgx as thiserror;
+}
+
+use codec::Encode;
+use error::Result;
+use itp_nonce_cache::{MutateNonce, Nonce};
+use itp_settings::node::{RUNTIME_SPEC_VERSION, RUNTIME_TRANSACTION_VERSION};
+use itp_types::OpaqueCall;
+use sp_core::{Pair, H256};
+use sp_runtime::{MultiSignature, OpaqueExtrinsic};
+use std::{sync::Arc, vec::Vec};
+use substrate_api_client::compose_extrinsic_offline;
+
+pub mod error;
+
+/// Create extrinsics from opaque calls
+///
+/// Also increases the nonce counter for each extrinsic that is created.
+pub trait CreateExtrinsics {
+	fn create_extrinsics(&self, calls: &[OpaqueCall]) -> Result<Vec<OpaqueExtrinsic>>;
+}
+
+/// Extrinsics factory
+pub struct ExtrinsicsFactory<Signer, NonceCache>
+where
+	Signer: Pair<Public = sp_core::ed25519::Public>,
+	Signer::Signature: Into<MultiSignature>,
+	NonceCache: MutateNonce,
+{
+	genesis_hash: H256,
+	signer: Signer,
+	nonce_cache: Arc<NonceCache>,
+}
+
+impl<Signer, NonceCache> ExtrinsicsFactory<Signer, NonceCache>
+where
+	Signer: Pair<Public = sp_core::ed25519::Public>,
+	Signer::Signature: Into<MultiSignature>,
+	NonceCache: MutateNonce,
+{
+	pub fn new(genesis_hash: H256, signer: Signer, nonce_cache: Arc<NonceCache>) -> Self {
+		ExtrinsicsFactory { genesis_hash, signer, nonce_cache }
+	}
+}
+
+impl<Signer, NonceCache> CreateExtrinsics for ExtrinsicsFactory<Signer, NonceCache>
+where
+	Signer: Pair<Public = sp_core::ed25519::Public>,
+	Signer::Signature: Into<MultiSignature>,
+	NonceCache: MutateNonce,
+{
+	fn create_extrinsics(&self, calls: &[OpaqueCall]) -> Result<Vec<OpaqueExtrinsic>> {
+		let mut nonce_lock = self.nonce_cache.load_for_mutation()?;
+		let mut nonce_value = nonce_lock.0;
+
+		let extrinsics_buffer: Vec<OpaqueExtrinsic> = calls
+			.iter()
+			.map(|call| {
+				let xt = compose_extrinsic_offline!(
+					self.signer.clone(),
+					call,
+					nonce_value,
+					Era::Immortal,
+					self.genesis_hash,
+					self.genesis_hash,
+					RUNTIME_SPEC_VERSION,
+					RUNTIME_TRANSACTION_VERSION
+				)
+				.encode();
+				nonce_value += 1;
+				xt
+			})
+			.map(|xt| {
+				OpaqueExtrinsic::from_bytes(&xt)
+					.expect("A previously encoded extrinsic has valid codec; qed.")
+			})
+			.collect();
+
+		*nonce_lock = Nonce(nonce_value);
```

Comment:
Retrieves the nonce write lock, creates all the extrinsics and increments the nonce value for each of them. Then at the end, write the nonce value back into the cache.
The reason we have a strict nonce lock here, is that we want to prevent any other concurrent process to get or write the nonce while we're incrementing it here.